[RISCV] Convert AVLs with vlenb to VLMAX where possible

lukel97 opened the pull request:
Given an AVL that's computed from vlenb, if it's equal to VLMAX then we can replace it with the VLMAX sentinel value.

The main motiviation is to be able to express an EVL of VLMAX in VP intrinsics whilst emitting vsetvli a0, zero, so that we can replace llvm.riscv.masked.strided.{load,store} with their VP counterparts.

This is done in RISCVVectorPeephole (previously RISCVFoldMasks, renamed to account for the fact that it no longer just folds masks) instead of SelectionDAG since there are multiple places places where VP nodes are lowered that would have need to have been handled.

This also avoids doing it in RISCVInsertVSETVLI as it's much harder to lookup the value of the AVL, and in RISCVVectorPeephole we can take advantage of DeadMachineInstrElim to remove any leftover PseudoReadVLENBs.

lukel97 committed:
[RISCV] Convert AVLs with vlenb to VLMAX where possible

Given an AVL that's computed from vlenb, if it's equal to VLMAX then we can replace it with the VLMAX sentinel value.

The main motiviation is to be able to express an EVL of VLMAX in VP intrinsics whilst emitting vsetvli a0, zero, so that we can replace llvm.riscv.masked.strided.{load,store} with their VP counterparts.

This is done in RISCVFoldMasks instead of SelectionDAG since there are multiple places places where VP nodes are lowered that would have need to have been handled.

This also avoids doing it in RISCVInsertVSETVLI as it's much harder to lookup the value of the AVL, and in RISCVFoldMasks we can take advantage of DeadMachineInstrElim to remove any leftover PseudoReadVLENBs.

lukel97 committed:
Fix typos

wangpc-pp commented on llvm/lib/Target/RISCV/RISCVFoldMasks.cpp:
```diff
@@ -62,6 +69,54 @@ char RISCVFoldMasks::ID = 0;
+// If an AVL is a VLENB that's possibly scaled to be equal to VLMAX, convert it
+// to the VLMAX sentinel value.
+bool RISCVFoldMasks::convertToVLMAX(MachineInstr &MI) const {
+  if (!RISCVII::hasVLOp(MI.getDesc().TSFlags) ||
```
Maybe we need a `isRVVPseudo` helper, I think it can be very useful.

lukel97 replied:
Yeah, I think RISCVInsertVSETVLI relies on hasSEWOp to check if an instruction is a pseudo.

lukel97 replied:
I think we could use `RISCVVPseudosTable::getPseudoInfo(Opc)` to check if it's a pseudo.

For here do we need to use it, or is it ok to just check for VL and SEW ops?

wangpc-pp replied:
Looking up table may be slow, I think it's OK to just check VL/SEW.

wangpc-pp commented:
I think we can rename `RISCVFoldMasks` to `PostISelRVVOptimization` or whatever.

lukel97 commented:
> I think we can rename `RISCVFoldMasks` to `PostISelRVVOptimization` or whatever.

Maybe something like `RISCVVectorPeephole`?

wangpc-pp commented:
> > I think we can rename `RISCVFoldMasks` to `PostISelRVVOptimization` or whatever.
> 
> Maybe something like `RISCVVectorPeephole`?

Make sense to me.

topperc commented on llvm/lib/Target/RISCV/RISCVFoldMasks.cpp:
```diff
@@ -62,6 +69,54 @@ char RISCVFoldMasks::ID = 0;
+  // Fixed-point value, denominator=8
+  unsigned LMULFixed = LMUL.second ? (8 / LMUL.first) : 8 * LMUL.first;
+  unsigned SEW =
+      1 << MI.getOperand(RISCVII::getSEWOpNum(MI.getDesc())).getImm();
```
Does this need to account for mask instructions having an SEW operand of 0?

lukel97 committed:
Precommit tests

lukel97 committed:
Handle SEW=1

lukel97 committed:
Rename RISCVFoldMasks -> RISCVVectorPeephole

lukel97 commented:
Ping

topperc commented on llvm/lib/Target/RISCV/RISCVVectorPeephole.cpp:
```diff
@@ -1,19 +1,30 @@
-//===- RISCVFoldMasks.cpp - MI Vector Pseudo Mask Peepholes ---------------===//
+//===- RISCVVectorPeephole.cpp - MI Vector Pseudo Peepholes --------------===//
```
Is this only 79 characters?

lukel97 committed:
Extend dividers to 80 columns

topperc commented on llvm/lib/Target/RISCV/RISCVVectorPeephole.cpp:
```diff
@@ -58,11 +70,62 @@ class RISCVFoldMasks : public MachineFunctionPass {
+  unsigned ScaleFixed = 8;
+  // Check if the VLENB was potentially scaled with slli/srli
+  if (Def->getOpcode() == RISCV::SLLI) {
+    ScaleFixed <<= Def->getOperand(2).getImm();
```
Do we need to check that the shift amount isn't out of bounds?

lukel97 committed:
Do scale in uint64_t and assert shift amount < 64

Also assert that 64 * LMUL / SEW > 0, so if the scale is shifted down to 0 then the equality check should always be false.

topperc reviewed (approved):
LGTM
